adding validation errors when the benchmarks are unsupported

emanuel-v-r opened the pull request:
Aims to fix #989, added a similar test to the one that is described in the issue.
After this PR, the summary will include validation errors and the output will look like this:
```
  Standard Output: 
// Benchmark BenchmarkAllCases.InvokeOnceVoid: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

// Benchmark BenchmarkAllCases.InvokeOnceTaskAsync: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

// Benchmark BenchmarkAllCases.InvokeOnceRefType: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

// Benchmark BenchmarkAllCases.InvokeOnceValueType: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

// Benchmark BenchmarkAllCases.InvokeOnceTaskOfTAsync: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

// Benchmark BenchmarkAllCases.InvokeOnceValueTaskOfT: Dry(Platform=X86, Toolchain=InProcessToolchain, InvocationCount=16, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=16, WarmupCount=1)
// cannot be run in-process. Validation errors:
//    * Job Dry, EnvironmentMode.Platform was run as X64 (X86 expected). Fix your test runner options.

No suitable benchmarks were found
```
Although I believe it fits the purpose of what is described in the original issue, the error is not that specific, if we want to have something more specific we will have to introduce a new method in the IToolChain interface as the existing one returns only a bool https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/IToolchain.cs#L16

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2148) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2148) before we can accept your contribution.<br/><br/>:x: emanuel-v-r [sign now](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2148)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/BenchmarkDotNet?pullRequest=2148) it.</sub>

YegorStepanov commented:
Yep, we need a specific error message in `summary.ValidationErrors`.

The problem is that the benchmarks are filtered by `Toolchain.IsSupported` before the validation executes.

```C#
var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, compositeLogger, resolver); //filtering by Toolchain.IsSupported
if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))
    return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };

var validationErrors = Validate(supportedBenchmarks, compositeLogger);
if (validationErrors.Any(validationError => validationError.IsCritical))
    return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };
```

---

> if we want to have something more specific we will have to introduce a new method in the IToolChain interface as the existing one returns only a bool https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/IToolchain.cs#L16

If we replace `ILogger logger` with `List<ValidationError> validationErrors`, we can do smth like this:
```C#
var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, resolver, out List<ValidationErrors> validationErrors);
validationErrors.AddRange(Validate(supportedBenchmarks, compositeLogger));

if (validationErrors.Any(validationError => validationError.IsCritical))
    return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };

if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))  // looks redundant, validators should check it
    return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };
```

---

For the maintainers: 
Validators are pretty bad at the moment, they need to be improved in some places. Someday I will write a "plan" issue. But the task is quite big.

emanuel-v-r commented:
> Yep, we need a specific error message in `summary.ValidationErrors`.
> 
> The problem is that the benchmarks are filtered by `Toolchain.IsSupported` before the validation executes.
> 
> ```cs
> var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, compositeLogger, resolver); //filtering by Toolchain.IsSupported
> if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))
>     return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };
> 
> var validationErrors = Validate(supportedBenchmarks, compositeLogger);
> if (validationErrors.Any(validationError => validationError.IsCritical))
>     return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };
> ```
> 
> > if we want to have something more specific we will have to introduce a new method in the IToolChain interface as the existing one returns only a bool https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/IToolchain.cs#L16
> 
> If we replace `ILogger logger` with `List<ValidationError> validationErrors`, we can do smth like this:
> 
> ```cs
> var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, resolver, out List<ValidationErrors> validationErrors);
> validationErrors.AddRange(Validate(supportedBenchmarks, compositeLogger));
> 
> if (validationErrors.Any(validationError => validationError.IsCritical))
>     return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };
> 
> if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))  // looks redundant, validators should check it
>     return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };
> ```
> 
> For the maintainers: Validators are pretty bad at the moment, they need to be improved in some places. Someday I will write a "plan" issue. But the task is quite big.



> Yep, we need a specific error message in `summary.ValidationErrors`.
> 
> The problem is that the benchmarks are filtered by `Toolchain.IsSupported` before the validation executes.
> 
> ```cs
> var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, compositeLogger, resolver); //filtering by Toolchain.IsSupported
> if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))
>     return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };
> 
> var validationErrors = Validate(supportedBenchmarks, compositeLogger);
> if (validationErrors.Any(validationError => validationError.IsCritical))
>     return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };
> ```
> 
> > if we want to have something more specific we will have to introduce a new method in the IToolChain interface as the existing one returns only a bool https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/IToolchain.cs#L16
> 
> If we replace `ILogger logger` with `List<ValidationError> validationErrors`, we can do smth like this:
> 
> ```cs
> var supportedBenchmarks = GetSupportedBenchmarks(benchmarkRunInfos, resolver, out List<ValidationErrors> validationErrors);
> validationErrors.AddRange(Validate(supportedBenchmarks, compositeLogger));
> 
> if (validationErrors.Any(validationError => validationError.IsCritical))
>     return new[] { Summary.ValidationFailed(title, resultsFolderPath, logFilePath, validationErrors) };
> 
> if (!supportedBenchmarks.Any(benchmarks => benchmarks.BenchmarksCases.Any()))  // looks redundant, validators should check it
>     return new[] { Summary.NothingToRun(title, resultsFolderPath, logFilePath) };
> ```
> 
> For the maintainers: Validators are pretty bad at the moment, they need to be improved in some places. Someday I will write a "plan" issue. But the task is quite big.

Thanks for the feedback, changed accordingly here https://github.com/dotnet/BenchmarkDotNet/pull/2148/commits/524dbf9f4ab3ae90a725001d9598f78301e2c258.
Notes:
* I haven't replaced logger because it's being used int the toolchain methods (ex: https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/InProcess/InProcessValidator.cs#L107), seems that the lib needs some more generic approach for validations/errors, so that we don't need to span "writeline" all over the place
* I have kept the original "IsSupported" untouched, so that we don't break any contract

Please let me know if that works for you

adamsitnik commented on tests/BenchmarkDotNet.IntegrationTests/InProcessTest.cs:
```diff
@@ -56,10 +57,28 @@ public InProcessTest(ITestOutputHelper output) : base(output)
+        {
+            var otherPlatform = IntPtr.Size == 8
+                ? Platform.X86
+                : Platform.X64;
```
this is really clever, thank you for adding the test! 👍

adamsitnik commented on tests/BenchmarkDotNet.IntegrationTests/InProcessTest.cs:
```diff
@@ -56,10 +57,28 @@ public InProcessTest(ITestOutputHelper output) : base(output)
+                : Platform.X64;
+
+            var otherPlatformConfig = new ManualConfig()
+                .With(Job.Dry.With(InProcessToolchain.Instance).WithInvocationCount(UnrollFactor).WithUnrollFactor(UnrollFactor).With(otherPlatform))
```
nit: `Job.Dry` already defines invocation count and unroll factor (1)

```suggestion
                .With(Job.Dry.With(InProcessToolchain.Instance).With(otherPlatform))
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/IToolchain.cs:
```diff
@@ -14,5 +17,6 @@ public interface IToolchain
         bool IsInProcess { get; }
 
         bool IsSupported(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
+        IEnumerable<ValidationError> CheckSupport(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
```
since this method returns `IEnumerable<ValidationError>`, it  could be called `Validate`. Moreover since we should change its responsibility to just return the errors (not report them), and don't pass `ILogger` instance to it:


```suggestion
        IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCase, IResolver resolver);
```

emanuel-v-r replied:
Actually the responsibility remains the same, but it returns an IEnumerable<ValidationError> instead of just a bool, so that the client knows specifically what happened.
Removing the logger means that we would have to create some generic approach for "printing" toolchain information, as the logger is used inside the toolchain implementations, in places like this https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/InProcess/InProcessValidator.cs#L107.
Renamed to Validate

adamsitnik commented on src/BenchmarkDotNet/Toolchains/IToolchain.cs:
```diff
@@ -14,5 +17,6 @@ public interface IToolchain
         bool IsInProcess { get; }
 
         bool IsSupported(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
```
Is `IsSupported` still needed? We should consider removing it.

emanuel-v-r replied:
Yes, still used

adamsitnik commented on src/BenchmarkDotNet/Toolchains/CoreRun/CoreRunToolchain.cs:
```diff
@@ -58,17 +61,27 @@ public class CoreRunToolchain : IToolchain
             {
-                logger.WriteLineError($"Provided CoreRun path does not exist, benchmark '{benchmark.DisplayInfo}' will not be executed. Please remember that BDN expects path to CoreRun.exe (corerun on Unix), not to Core_Root folder.");
-                return false;
+                var validationError = new ValidationError(true, $"Provided CoreRun path does not exist, benchmark '{benchmark.DisplayInfo}' will not be executed. Please remember that BDN expects path to CoreRun.exe (corerun on Unix), not to Core_Root folder.", benchmark);
+
+                logger.WriteLineError(validationError.Message);
+
+                return new List<ValidationError> { validationError };
```
we can take advantage of yield return/break and simplify the code a bit:

```suggestion
                yield return new ValidationError(true, $"Provided CoreRun path does not exist, benchmark '{benchmark.DisplayInfo}' will not be executed. Please remember that BDN expects path to CoreRun.exe (corerun on Unix), not to Core_Root folder.", benchmark);
                yield break;
```

emanuel-v-r replied:
Do you think it's worth it? Seems that it requires the same amount of code. I tend to avoid the yield keyword for finite collections, as it uses a state machine under the hood.

adamsitnik replied:
I do. It simplifies the code + by returning `List` we return a mutable collection, which can be misused by some callers:

```cs
IEnumerable<ValidationError> collection = Validate(benchmark, resolver);
((List<ValidationError>)collection).Add(new ...)
```

In places where we know the size of the collection and it's fixed we can use arrays instead of `yield` or even `ImmutableArray` to make sure that the content never changes.

adamsitnik commented on src/BenchmarkDotNet/Toolchains/CoreRun/CoreRunToolchain.cs:
```diff
@@ -58,17 +61,27 @@ public class CoreRunToolchain : IToolchain
             if (Toolchain.InvalidCliPath(CustomDotNetCliPath?.FullName, benchmark, logger))
-                return false;
+            {
+                var validationError = new ValidationError(true, "Invalid Cli Path", benchmark);
+
+                return new List<ValidationError> { validationError };
```
```suggestion
                yield return new ValidationError(true, "Invalid Cli Path", benchmark);
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/CoreRun/CoreRunToolchain.cs:
```diff
@@ -58,17 +61,27 @@ public class CoreRunToolchain : IToolchain
+            }
 
-            return true;
+            return Enumerable.Empty<ValidationError>();
```
```suggestion
```

adamsitnik commented on src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:
```diff
@@ -531,13 +532,26 @@ private static BenchmarkReport RunCore(BenchmarkCase benchmarkCase, BenchmarkId
-        private static BenchmarkRunInfo[] GetSupportedBenchmarks(BenchmarkRunInfo[] benchmarkRunInfos, ILogger logger, IResolver resolver)
-            => benchmarkRunInfos.Select(info => new BenchmarkRunInfo(
-                    info.BenchmarksCases.Where(benchmark => benchmark.GetToolchain().IsSupported(benchmark, logger, resolver)).ToArray(),
+        private static BenchmarkRunInfo[] GetSupportedBenchmarks(BenchmarkRunInfo[] benchmarkRunInfos, ILogger logger, IResolver resolver, out List<ValidationError> validationErrors)
```
inside `BenchmarkRunnerClean` it's OK to use `ValueTuples`, we could use them here as well:

```suggestion
        private static (BenchmarkRunInfo[] benchmarks, IReadOnlyList<ValidationError> validationErrors) GetSupportedBenchmarks(BenchmarkRunInfo[] benchmarkRunInfos, ILogger logger, IResolver resolver)
```

adamsitnik reviewed (commented):
@emanuel-v-r big thanks for your contribution! PTAL at my comments

YegorStepanov commented:
I guess this test won't contain `validationError`:
```C#
[Fact]
public void BenchmarkDifferentPlatformReturnsValidationError()
{
    var сonfig = new ManualConfig()
        .With(Job.Dry.With(InProcessToolchain.Instance).With(Platform.X86))
        .With(Job.Dry.With(InProcessToolchain.Instance).With(Platform.X64))
        .With(new OutputLogger(Output))
        .With(DefaultColumnProviders.Instance);

    var runInfo = BenchmarkConverter.TypeToBenchmarks(typeof(BenchmarkAllCases), сonfig);
    var summary = BenchmarkRunner.Run(runInfo);

    Assert.NotEmpty(summary.ValidationErrors);
}
```
Also, we need to add a specific reason to the validation errors.

To fix it we need change interface to it 
```diff
public interface IToolchain                                                            
{                                                                                     
    [PublicAPI] string Name { get; }                                                   
    IGenerator Generator { get; }                                                      
    IBuilder Builder { get; }                                                          
    IExecutor Executor { get; }                                                        
    bool IsInProcess { get; }                                                          
                                                                                      
-   bool IsSupported(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
+   bool IsSupported(BenchmarkCase benchmarkCase, IResolver resolver, List<ValidationError> validationErrors);
}                                                                                     
```

and do it for each toolchain:
```diff
- logger.WriteLineError("The Roslyn toolchain is only supported on .NET Framework");
+ validationErrors.Add(new ValidationError(false, "The Roslyn toolchain is only supported on .NET Framework"));
```

But:
1) We need approval from the maintainers to do it.
2) Refactoring is desirable here, but it requires an understanding of the project's code
2.1) For example: change `List<ValidationError>` to the add-only collection `ValidationErrors` in the whole project.

emanuel-v-r commented:
> I guess this test won't contain `validationError`:
> 
> ```cs
> [Fact]
> public void BenchmarkDifferentPlatformReturnsValidationError()
> {
>     var сonfig = new ManualConfig()
>         .With(Job.Dry.With(InProcessToolchain.Instance).With(Platform.X86))
>         .With(Job.Dry.With(InProcessToolchain.Instance).With(Platform.X64))
>         .With(new OutputLogger(Output))
>         .With(DefaultColumnProviders.Instance);
> 
>     var runInfo = BenchmarkConverter.TypeToBenchmarks(typeof(BenchmarkAllCases), сonfig);
>     var summary = BenchmarkRunner.Run(runInfo);
> 
>     Assert.NotEmpty(summary.ValidationErrors);
> }
> ```
> 
> Also, we need to add a specific reason to the validation errors.
> 
> To fix it we need change interface to it
> 
> ```diff
> public interface IToolchain                                                            
> {                                                                                     
>     [PublicAPI] string Name { get; }                                                   
>     IGenerator Generator { get; }                                                      
>     IBuilder Builder { get; }                                                          
>     IExecutor Executor { get; }                                                        
>     bool IsInProcess { get; }                                                          
>                                                                                       
> -   bool IsSupported(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
> +   bool IsSupported(BenchmarkCase benchmarkCase, IResolver resolver, List<ValidationError> validationErrors);
> }                                                                                     
> ```
> 
> and do it for each toolchain:
> 
> ```diff
> - logger.WriteLineError("The Roslyn toolchain is only supported on .NET Framework");
> + validationErrors.Add(new ValidationError(false, "The Roslyn toolchain is only supported on .NET Framework"));
> ```
> 
> But:
> 
> 1. We need approval from the maintainers to do it.
> 2. Refactoring is desirable here, but it requires an understanding of the project's code
>    2.1) For example: change `List<ValidationError>` to the add-only collection `ValidationErrors` in the whole project.

I was not able to replicate the issue using your test but I do understand what the issue is, nice catch thank you!
This commit should do the trick https://github.com/dotnet/BenchmarkDotNet/pull/2148/commits/c45077d5b04a0751f990559c5f39ee2166166dff.
Also @adamsitnik thanks for your feedback, I addressed most of your comments, but I kept open some of them that I still have some doubts.
More specifically this one https://github.com/dotnet/BenchmarkDotNet/pull/2148#discussion_r995790537.
Seems that we all agree that this validate method should not log/print anything, and therefore the logger should be removed.
The problem is that inside the toolchain implementations, we do have some logging/printing, and there's no existing generic approach for such error handling.
My suggestion is that we create a separated issue for handling this, as it could take some effort.
Anyway the changes that I have done in this PR will not affect such behavior.

YegorStepanov commented on src/BenchmarkDotNet/Loggers/LoggerExtensions.cs:
```diff
@@ -20,7 +21,9 @@ public static class LoggerExtensions
-        public static void Write(this ILogger logger, string text) =>  logger.Write(LogKind.Default, text);
+        public static void Write(this ILogger logger, string text) => logger.Write(LogKind.Default, text);
+
+        public static void WriteLineValidationError(this ILogger logger, ValidationError validationError) => logger.WriteLineError(validationError.Message);
```
It's redundant since it will only be used in one place (when you remove `Logger` from `Validate`), it's best not to duplicate methods.

YegorStepanov replied:
@emanuel-v-r When the `Toolchain.InvalidCliPath(...) == false` you don't return actual error, so it's in the scope 😊

The new version of BDN should be out soon (1-2 weeks).
But it looks unfinished for the release (IMO)

```С#
public interface IToolchain
{
    bool IsSupported(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
    IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCase, ILogger logger, IResolver resolver);
}
```

```C#
var validationError = new ValidationError(true, "...", benchmarkCase);
logger.WriteLineValidationError(validationError);
return validationError.AsEnumerable();
```

And it would be cool, if you add a few tests that hits, for example, `Toolchain.InvalidCliPath` path.

emanuel-v-r replied:
I am talking about removing logger dependencies from the `Toolchain` implementations.
As you mentioned `Toolchain.InvalidCliPath` is in scope, and it was addressed in scope of your other comment https://github.com/dotnet/BenchmarkDotNet/pull/2148#discussion_r995925947.
Here's the commit https://github.com/dotnet/BenchmarkDotNet/pull/2148/commits/40ae83c481bca11bc5855852ee7e203eb87f7c7a

YegorStepanov commented on src/BenchmarkDotNet/Toolchains/CsProj/CsProjClassicNetToolchain.cs:
```diff
@@ -33,21 +36,34 @@ private CsProjClassicNetToolchain(string targetFrameworkMoniker, string name, st
             if (InvalidCliPath(customDotNetCliPath: null, benchmarkCase, logger))
-                return false;
+            {
+                var validationError = new ValidationError(true, $"InvalidCliPath, benchmark '{benchmarkCase.DisplayInfo}' will not be executed", benchmarkCase);
```
Do not use fast exit, collect all errors and return them.

It's better to display as much as possible errors.
If the user is not on windows AND the cli path is invalid, it displays only the first message.

emanuel-v-r replied:
I am not sure if we want to check all the errors, or just fail fast here, anyway I kep the existing behavior, not sure if we want to change that.

YegorStepanov replied:
You can open the validators folder, all of them may display multiple errors.

emanuel-v-r replied:
> You can open the validators folder, all of them may display multiple errors.

I mean toolchains errors, currently it only prints one error and exits as you can see in the previous code for "IsSupported" method
I am ok to change it as well, but I am trying to be objective here, and change only the necessary for what is described in the issue, and try not have side effects
Maybe @adamsitnik can provide his opinion here.

adamsitnik replied:
I like the idea of returning all errors instead the first one. 👍 

Currently if there are two or more errors:

1. The users tries to run the benchmarks, gets a single error and fixes it.
2. The users tries to run the benchmarks, gets another error and fixes it.
3. The user actually runs the benchmarks.

If we provide all errors at once we can reduce the number of steps needed and hence improve the UX.

YegorStepanov commented on src/BenchmarkDotNet/Validators/ValidationError.cs:
```diff
@@ -50,6 +51,8 @@ public override int GetHashCode()
             }
         }
 
+        internal IEnumerable<ValidationError> AsEnumerable() => new List<ValidationError>() { this };
```
If you remove "fast-exit" (as I suggested above), it will be redundant.

adamsitnik replied:
this method should not be needed as we can just use `yield` everywhere. Also, if we really needed that it should be returning an array (with the list the caller can cast it to list and add some elements) + array is cheaper than list `new [1] { this }` 

```suggestion
```

YegorStepanov commented on src/BenchmarkDotNet/Toolchains/CsProj/CsProjClassicNetToolchain.cs:
```diff
@@ -33,21 +36,34 @@ private CsProjClassicNetToolchain(string targetFrameworkMoniker, string name, st
+                return validationError.AsEnumerable();
             }
 
             if (InvalidCliPath(customDotNetCliPath: null, benchmarkCase, logger))
```
You should refactor InvalidCLIPath and return validation errors from it. After it you will can delete logger

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
+        public override IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCase, IResolver resolver)
         {
-            if (!base.IsSupported(benchmarkCase, logger, resolver))
+            var baseValidations = base.Validate(benchmarkCase, resolver);
+
+            if (baseValidations.Any())
```
```suggestion
            foreach (var validationError in base.Validate(benchmarkCase, resolver))
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
+            if (baseValidations.Any())
             {
-                return false;
+                return baseValidations;
```
```suggestion
                yield return validationError;
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             {
-                logger.WriteLineError("The MonoAOT toolchain requires the Runtime property to be configured explicitly to an instance of MonoRuntime class");
-                return false;
+                return
+                    new ValidationError(true, "The MonoAOT toolchain requires the Runtime property to be configured explicitly to an instance of MonoRuntime class", benchmarkCase)
+                    .AsEnumerable();
```
```suggestion
                yield return
                    new ValidationError(true, "The MonoAOT toolchain requires the Runtime property to be configured explicitly to an instance of MonoRuntime class", benchmarkCase);
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             {
-                logger.WriteLineError($"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.");
-                return false;
+                return
+                    new ValidationError(true, $"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.", benchmarkCase)
+                    .AsEnumerable();
```
```suggestion
                yield return
                    new ValidationError(true, $"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.", benchmarkCase);
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             {
-                logger.WriteLineError("The MonoAOT toolchain does not allow to rebuild source project, so defining custom build configuration makes no sense");
-                return false;
+                return
+                    new ValidationError(true, "The MonoAOT toolchain does not allow to rebuild source project, so defining custom build configuration makes no sense", benchmarkCase)
+                    .AsEnumerable();
```
```suggestion
                yield return
                    new ValidationError(true, "The MonoAOT toolchain does not allow to rebuild source project, so defining custom build configuration makes no sense", benchmarkCase);
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             {
-                logger.WriteLineError("The MonoAOT toolchain does not allow specifying NuGet package dependencies");
-                return false;
+                var validationError = new ValidationError(true, "The MonoAOT toolchain does not allow specifying NuGet package dependencies", benchmarkCase);
+
+                return validationError.AsEnumerable();
```
```suggestion
                yield return new ValidationError(true, "The MonoAOT toolchain does not allow specifying NuGet package dependencies", benchmarkCase);
```

adamsitnik commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +21,45 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             }
 
-            return true;
+            return Enumerable.Empty<ValidationError>();
```
```suggestion
```

adamsitnik reviewed (commented):
Overall, it looks very good, but it would be great if you could change the validator behavior to return all known errors when possible. I know that the original issue did not mention it, but since you are improving the error handling, why not make it even better when we can?

Thank you @emanuel-v-r !

YegorStepanov commented on src/BenchmarkDotNet/Toolchains/Toolchain.cs:
```diff
@@ -29,39 +29,34 @@ public Toolchain(string name, IGenerator generator, IBuilder builder, IExecutor
         }
 
-        internal static bool InvalidCliPath(string customDotNetCliPath, BenchmarkCase benchmarkCase, out ValidationError validationError)
+        internal static bool IsCliPathInvalid(string customDotNetCliPath, BenchmarkCase benchmarkCase, out ValidationError validationError)
```
Hmm, `IsCliPathInvalid` returns one of two validation errors always. You can leave the current version.

emanuel-v-r commented:
> Overall, it looks very good, but it would be great if you could change the validator behavior to return all known errors when possible. I know that the original issue did not mention it, but since you are improving the error handling, why not make it even better when we can?
> 
> Thank you @emanuel-v-r !

Thank you Adam, applied your suggestions, and also changed the validate method in the other toolchains so that it returns multiple erros.

YegorStepanov commented on src/BenchmarkDotNet/Running/BenchmarkRunnerDirty.cs:
```diff
@@ -136,7 +136,7 @@ private static Summary RunWithExceptionHandling(Func<Summary> run)
             {
                 ConsoleLogger.Default.WriteLineError(e.Message);
-                return Summary.NothingToRun(e.Message, string.Empty, string.Empty);
+                return Summary.ValidationFailed(e.Message, string.Empty, string.Empty);
```
Looks a little weird here. Maybe find a better name?

emanuel-v-r replied:
Any suggestion? Maybe just Summary.Failed?

YegorStepanov commented on src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:
```diff
@@ -531,14 +532,27 @@ private static BenchmarkReport RunCore(BenchmarkCase benchmarkCase, BenchmarkId
+            var benchmarksRunInfo = benchmarkRunInfos.Select(info => new BenchmarkRunInfo(
+                    info.BenchmarksCases.Where(benchmark =>
+                    {
+                        var benchmarkValidationErrors = benchmark.GetToolchain().Validate(benchmark, resolver);
+
+                        validationErrors.AddRange(benchmarkValidationErrors);
+
+                        return !benchmarkValidationErrors.Any();
```
"the more nested the type, the shorter its name" 

Bug: it will execute `Validation` twice.


```suggestion
                        var errors = benchmark.GetToolchain().Validate(benchmark, resolver).ToArray();
                        validationErrors.AddRange(errors);
                        return !errors.Any();
```

YegorStepanov commented on src/BenchmarkDotNet/Toolchains/Mono/MonoAotToolchain.cs:
```diff
@@ -18,39 +19,36 @@ public MonoAotToolchain() : base("MonoAot", new Generator(), new MonoAotBuilder(
             {
-                logger.WriteLineError($"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.");
-                return false;
+                yield return
+                    new ValidationError(true, $"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.", benchmarkCase);
```
```suggestion
                yield return new ValidationError(true, $"The MonoBclPath provided for MonoAOT toolchain: {monoRuntime.MonoBclPath} does NOT exist.", benchmarkCase);
```

emanuel-v-r replied:
It was actually me, I do prefer this way, to avoid such a long line.
My IDE does not complain about it as well...

YegorStepanov replied:
It not critical, but the project uses 2 types of formatting for validation errors:
```C#
//#1
yield return new ValidationError(true, "The MonoAOT toolchain requires the Runtime property to be configured explicitly to an instance of MonoRuntime class", benchmarkCase);
//#2
yield return new ValidationError(true,
    "The MonoAOT toolchain requires the Runtime property to be configured explicitly to an instance of MonoRuntime class", 
    benchmarkCase);
```

The perfectionist is crying 😊
![image](https://user-images.githubusercontent.com/19392641/196255616-cbb34d12-6858-4e84-a3ff-4e2b482c5b50.png)

YegorStepanov replied:
> Than this: ![image](https://user-images.githubusercontent.com/40024962/196255815-0b10b40e-601e-4843-98b1-da532e717532.png)

@emanuel-v-r FYI:
Visual Studio split it by 2 lines due to [soft wrap mode](https://learn.microsoft.com/en-gb/visualstudio/ide/reference/how-to-manage-word-wrap-in-the-editor?view=vs-2022
). You can toggle it

adamsitnik commented on src/BenchmarkDotNet/Toolchains/MonoWasm/WasmToolchain.cs:
```diff
@@ -1,40 +1,46 @@
 {
     [PublicAPI]
-    public class WasmToolChain : Toolchain
+    public class WasmToolchain : Toolchain
```
Renaming public types is considered to be a breaking change and we avoid doing that if we can. However, this particular type is most likely not used directly by anyone, so it's OK-ish ;)

adamsitnik commented on src/BenchmarkDotNet/Validators/ValidationError.cs:
```diff
@@ -50,6 +51,7 @@ public override int GetHashCode()
             }
         }
 
+
```
nit: redundant empty lone
